src/qinfer/distributions.py

```diff
@@ -321,6 +322,76 @@ def sample(self, n=1):
         ), len(cumsum_weights) - 1)]
 
     ## MOMENT FUNCTIONS ##
+    @staticmethod
+    def particle_mean(weights, locations):
+        r"""
+        Returns the mean of a function :math:`f` over model
+        parameters.
+
+        :param numpy.ndarray weights: Weights of each particle.
+        :param numpy.ndarray locations: Locations of each
+            particle.
+        :param callable fn: Function of model parameters to
+            take the mean of. If `None`, the identity function
+            is assumed.
+        """
+        # We need the particle index to be the rightmost index, so that
+        # the two arrays align on the particle index as opposed to the
+        # modelparam index.
+        return np.sum(weights * locations.transpose([1, 0]), axis=1)
+
+    @classmethod
+    def particle_covariance_mtx(cls, weights, locations):
+        """
+        Returns an estimate of the covariance of a distribution
+        represented by a given set of SMC particle.
+
+        :param weights: An array containing the weights of each
+            particle.
+        :param location: An array containing the locations of
+            each particle.
+        :rtype: :class:`numpy.ndarray`, shape
+            ``(n_modelparams, n_modelparams)``.
+        :returns: An array containing the estimated covariance matrix.
+        """
+        # TODO: add shapes to docstring.
+
+        # Find the mean model vector, shape (n_modelparams, ).
+        mu = cls.particle_mean(weights, locations)
```

Comment:
I think the current implementation leaves open a subtle edge case, as @ihincks says, in that a subclass could naively override ``particle_mean`` without also overriding ``particle_covariance_mtx``, such that the ``mu`` here would then be incorrect for the calculation in the rest of the method. Using ``ParticleDistribution.particle_mean`` explicitly would at least somewhat help by ensuring that even if the covariance is not consistent with the circular mean, it at least makes sense independently. Sadly, I don't know of a way in Python to force subclasses to override either neither or both methods, so maybe we should note in documentation that subclasses can't override just one?